sql/core/src/test/scala/org/apache/spark/sql/DataFrameTransposeSuite.scala

```diff
@@ -0,0 +1,173 @@
+      Row("salary", 2000.0, 1000.0) :: Nil
+    )
+  }
+
+  test("transpose with specified index column") {
+    checkAnswer(
+      salary.transpose($"salary"),
+      Row("personId", 1, 0) :: Nil
+    )
+  }
+
+  //
+  // Test cases: API behavior
+  //
+
+  test("enforce least common type for non-index columns") {
+    val df = Seq(("x", 1, 10.0), ("y", 2, 20.0)).toDF("name", "id", "value")
+    val transposedDf = df.transpose()
+    checkAnswer(
+      transposedDf,
+      Row("id", 1.0, 2.0) :: Row("value", 10.0, 20.0) :: Nil
+    )
+    // (id,IntegerType) -> (x,DoubleType)
+    // (value,DoubleType) -> (y,DoubleType)
+    assertResult(DoubleType)(transposedDf.schema("x").dataType)
+    assertResult(DoubleType)(transposedDf.schema("y").dataType)
+
+    val exception = intercept[AnalysisException] {
+      person.transpose()
+    }
+    assert(exception.getMessage.contains(
+      "[TRANSPOSE_NO_LEAST_COMMON_TYPE] Transpose requires non-index columns " +
+        "to share a least common type"))
+  }
+
+  test("enforce ascending order based on index column values for transposed columns") {
+    val transposedDf = person.transpose($"name")
+    checkAnswer(
+      transposedDf,
+      Row("id", 1, 0) :: Row("age", 20, 30)  :: Nil
+    )
+    // mike, jim -> jim, mike
+    assertResult(Array("key", "jim", "mike"))(transposedDf.columns)
+  }
+
+  test("enforce AtomicType Attribute for index column values") {
+    val exceptionAtomic = intercept[AnalysisException] {
+      complexData.transpose($"m")  // (m,MapType(StringType,IntegerType,false))
+    }
+    assert(exceptionAtomic.getMessage.contains(
+      "[TRANSPOSE_INVALID_INDEX_COLUMN] Invalid index column for TRANSPOSE because:" +
+        " Index column must be of atomic type, but found"))
+
+    val exceptionAttribute = intercept[AnalysisException] {
+      // (s,StructType(StructField(key,IntegerType,false),StructField(value,StringType,true)))
+      complexData.transpose($"s.key")
+    }
+    assert(exceptionAttribute.getMessage.contains(
+      "[TRANSPOSE_INVALID_INDEX_COLUMN] Invalid index column for TRANSPOSE because:" +
+        " Index column must be an atomic attribute"))
+  }
+
+  test("enforce transpose max values") {
+    spark.conf.set(SQLConf.DATAFRAME_TRANSPOSE_MAX_VALUES.key, 1)
+    val exception = intercept[AnalysisException](
+      person.transpose($"name")
+    )
+    assert(exception.getMessage.contains(
+      "[TRANSPOSE_EXCEED_ROW_LIMIT] Number of rows exceeds the allowed limit of"))
+    spark.conf.set(SQLConf.DATAFRAME_TRANSPOSE_MAX_VALUES.key,
+      SQLConf.DATAFRAME_TRANSPOSE_MAX_VALUES.defaultValue.get)
+  }
+
+  //
+  // Test cases: special frame
+  //
+
+  test("transpose empty frame w. column names") {
+    val schema = StructType(Seq(
+      StructField("id", IntegerType),
+      StructField("name", StringType)
+    ))
+    val emptyDF = spark.createDataFrame(spark.sparkContext.emptyRDD[Row], schema)
+    checkAnswer(
+      emptyDF.transpose(),
```

Comment:
can we also check the schema of the transposed df?

Reply:
Added check for schema.